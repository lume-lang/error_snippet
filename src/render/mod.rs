use crate::Diagnostic;

pub mod graphical;

pub use graphical::*;

/// Represents a wrapper around a standard formatter.
pub struct Formatter<'a> {
    inner: &'a mut dyn std::fmt::Write,
}

impl std::fmt::Write for Formatter<'_> {
    fn write_str(&mut self, s: &str) -> std::fmt::Result {
        self.inner.write_str(s)
    }
}

/// Defines a trait for rendering diagnostics to a formatter.
pub trait Renderer {
    /// Renders the diagnostic to a string buffer.
    fn render(&mut self, diagnostic: &dyn Diagnostic) -> Result<String, std::fmt::Error> {
        let mut buffer = String::new();
        let mut formatter = Formatter { inner: &mut buffer };

        self.render_fmt(&mut formatter, diagnostic)?;

        Ok(buffer)
    }

    /// Renders the diagnostic to the standard output buffer.
    fn render_stderr(&mut self, diagnostic: &dyn Diagnostic) -> std::fmt::Result {
        eprint!("{}", self.render(diagnostic)?);

        Ok(())
    }

    /// Renders the diagnostic to the given formatter.
    fn render_fmt(&mut self, f: &mut Formatter, diagnostic: &dyn Diagnostic) -> std::fmt::Result;
}
