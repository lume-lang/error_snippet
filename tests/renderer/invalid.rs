use std::sync::Arc;

use error_snippet::*;
use insta::assert_snapshot;

use crate::render;

#[test]
fn label_range_outside_range() {
    let source = Arc::new(NamedSource::new("src/test.lm", "abc"));

    let message = SimpleDiagnostic::new("failed to read file").with_label(Label::new(
        Some(source),
        256..280,
        "label_range_outside_range",
    ));

    assert_snapshot!(render(message));
}

#[test]
fn label_end_outside_range() {
    let source = Arc::new(NamedSource::new("src/test.lm", "abc"));

    let message = SimpleDiagnostic::new("failed to read file").with_label(Label::new(
        Some(source),
        0..50,
        "label_end_outside_range",
    ));

    assert_snapshot!(render(message));
}

#[test]
fn label_range_negative_length() {
    let source = Arc::new(NamedSource::new("src/test.lm", "abc"));

    let message = SimpleDiagnostic::new("failed to read file").with_label(Label::new(
        Some(source),
        3..1,
        "label_range_negative_length",
    ));

    assert_snapshot!(render(message));
}
