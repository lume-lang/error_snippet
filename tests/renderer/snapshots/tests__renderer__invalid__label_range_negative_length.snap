---
source: tests/renderer/invalid.rs
expression: render(message)
---
× error: failed to read file
   ╭─[src/test.lm:1:4]
 1 │ abc
   ∶   label_range_negative_length
   ╰──
